mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,444 @@
+            target_peer,
+            expires_at: Utc::now().timestamp() + expires_in_seconds,
+            is_a_reply,
+        };
+
+        let signature = try_s!(keypair.sign(&try_s!(data.encode())));
+
+        Ok(Self { signature, data })
+    }
+
+    pub(crate) fn is_received_message_valid(&self, my_peer_id: PeerId) -> bool {
+        let now = Utc::now().timestamp();
+        if now > self.data.expires_at {
+            log::debug!(
+                "Healthcheck message is expired. Current time in UTC: {now}, healthcheck `expires_at` in UTC: {}",
+                self.data.expires_at
+            );
+            return false;
+        }
+
+        if self.data.target_peer != my_peer_id {
+            log::debug!(
+                "`target_peer` doesn't match with our peer address. Our address: '{}', healthcheck `target_peer`: '{}'.",
+                my_peer_id,
+                self.data.target_peer
+            );
+            return false;
+        }
+
+        let Ok(public_key) = Libp2pPublic::try_decode_protobuf(&self.data.sender_public_key) else {
+            log::debug!("Couldn't decode public key from the healthcheck message.");
+
+            return false
+        };
+
+        if self.data.sender_peer != public_key.to_peer_id() {
+            log::debug!("`sender_peer` and `sender_public_key` doesn't belong each other.");
+
+            return false;
+        }
+
+        let Ok(encoded_message) = self.data.encode() else {
+            log::debug!("Couldn't encode healthcheck data.");
+            return false
+        };
+
+        let res = public_key.verify(&encoded_message, &self.signature);
+
+        if !res {
+            log::debug!("Healthcheck isn't signed correctly.");
+        }
+
+        res
+    }
+
+    #[inline]
+    pub(crate) fn encode(&self) -> Result<Vec<u8>, rmp_serde::encode::Error> { encode_message(self) }
+
+    #[inline]
+    pub(crate) fn decode(bytes: &[u8]) -> Result<Self, rmp_serde::decode::Error> { decode_message(bytes) }
+
+    #[inline]
+    pub(crate) fn should_reply(&self) -> bool { !self.data.is_a_reply }
+
+    #[inline]
+    pub(crate) fn sender_peer(&self) -> PeerId { self.data.sender_peer }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[cfg_attr(any(test, target_arch = "wasm32"), derive(PartialEq))]
+struct HealthcheckData {
+    #[serde(deserialize_with = "deserialize_peer_id", serialize_with = "serialize_peer_id")]
+    sender_peer: PeerId,
+    #[serde(deserialize_with = "deserialize_bytes")]
+    sender_public_key: Vec<u8>,
+    #[serde(deserialize_with = "deserialize_peer_id", serialize_with = "serialize_peer_id")]
+    target_peer: PeerId,
+    expires_at: i64,
+    is_a_reply: bool,
+}
+
+impl HealthcheckData {
+    #[inline]
+    fn encode(&self) -> Result<Vec<u8>, rmp_serde::encode::Error> { encode_message(self) }
+}
+
+#[inline]
+pub fn peer_healthcheck_topic(peer_id: &PeerId) -> String {
+    pub_sub_topic(PEER_HEALTHCHECK_PREFIX, &peer_id.to_string())
+}
```

Comment:
Why do we need the `peer_id` in the topic if we don't use it here https://github.com/KomodoPlatform/komodo-defi-framework/blob/13fe925e3c62c95b980de8ec279119a8593e0f70/mm2src/mm2_main/src/lp_network.rs#L218-L220

Reply:
Nevermind, we need the peer to only subscribe to it's topic to not receive all messages :)